Introduction to Free Monads/article.md

```diff
@@ -0,0 +1,892 @@
+    <> printState x s'
+```
+
+If we run `someComputation` defined above through this interpreter, for example by invoking `printState someComputation 1`, we'll get the following output:
+
+```
+state change 1 -> 1
+state change 1 -> 2
+pure ((),2)
+```
+
+The first line of output corresponds to `get`.
+While it doesn't modify the state, it's still an action, so the `state change` line is still printed.
+The second line corresponds to `put $ i + 1`, which naturally changes the state to `2`.
+The third line corresponds to `pure ()`, but it also outputs the end state.
+
+Let's summarize what we've learned so far.
+We can take any usual base functor for some monad, and get a `Monad` instance "for free".
+There's no free lunch here though, so we need to define the semantics of this monad separately by writing an interpreter (or several if we need to).
+
+### List as a free monad
+
+Not all free monads behave exactly the same as their regular counterparts.
+For instance, you might know that the list monad encodes non-determinism[^non-determinism].
+However, if we derive a free monad for lists, the behavior is slightly different.
+Consider:
+
+[^non-determinism]: If the statement about list monad encoding non-determinism is surprising, consider that one could use a list to represent all possible outcomes of some event, and then handle those one by one.
+    For further reading on this topic, refer to the [School of Haskell article on the list monad][school-of-haskell] or to the [Haskell wikibook section on understanding monads][haskell-wikibook]
+
+    [school-of-haskell]: https://www.schoolofhaskell.com/school/starting-with-haskell/basics-of-haskell/13-the-list-monad#non-deterministic-computations
+    [haskell-wikibook]: https://en.wikibooks.org/wiki/Haskell/Understanding_monads/List
+
+```haskell
+listComputation :: Free [] Int
+listComputation = do
+  x <- liftF [1, 2, 3]
+  y <- liftF [10, 20]
+  z <- liftF [100, 200]
+  pure $ x+y+z
+
+printFreeList :: Show a => Free [] a -> String
+printFreeList (Pure x) = show x
+printFreeList (Free f) = "["
+  <> intercalate "," (printFreeList <$> f)
+  <> "]"
+```
+
+if we run `printFreeList listComputation`, we will get
+
+```haskell
+[[[111,211],[121,221]],[[112,212],[122,222]],[[113,213],[123,223]]]
+```
+
+Notice how it gets us what is essentially nested lists, unlike the regular list monad.
+Notice also that it's essentially a rose tree.
+Of course, we can get the standard list monad behavior by concatenating the nested lists, but we might not necessarily want to do that.
+
+In general, we can restore any proper monad behavior from a free monad, since it doesn't define any semantics beyond those necessary for any monad.
+`free` provides a function for that:
+
+```haskell
+foldFree :: Monad m => (forall x. f x -> m x) -> Free f a -> m a
+```
+
+Given a function[^natural-transformation] converting `f x` into some monad `m x` for any `x`, we can convert `Free f a` into `m a`.
```

Comment:
I think there should be a comma before "for any x"?

Reply:
Uh... Me not being a native speaker can make cases like this a bit hard.

So, what I'm _trying_ to say here is this: given a function `f : ∀ x. f x → m x`, we can get `g : ∀ a. Free f a -> m a`. Hence, "for any `x`" isn't a qualifier, it's what's defining the "given function". With a comma, it sounds to me more like "for any `x`" qualifies the second sentence, i.e. "we can convert", which isn't at all the intended meaning.

Reply:
Mhm. Yeah, then it is not necessary.